homeassistant/components/withings/common.py

```diff
@@ -283,61 +797,308 @@ async def update_sleep_summary(self) -> SleepGetSummaryResponse:
+        attribute.platform, const.DOMAIN, unique_id
+    )
+
+    if entity_id is None:
+        _LOGGER.error("Cannot find entity id for unique_id: %s", unique_id)
+        return None
+
+    return entity_id
+
+
+class BaseWithingsSensor(Entity):
+    """Base class for withings sensors."""
+
+    def __init__(self, data_manager: DataManager, attribute: WithingsAttribute) -> None:
+        """Initialize the Withings sensor."""
+        self._data_manager = data_manager
+        self._attribute = attribute
+        self._profile = self._data_manager.profile
+        self._user_id = self._data_manager.user_id
+        self._name = f"Withings {self._attribute.measurement.value} {self._profile}"
+        self._unique_id = get_attribute_unique_id(self._attribute, self._user_id)
+        self._state_data: Optional[Any] = None
+
+    @property
+    def should_poll(self) -> bool:
+        """Return False to indicate HA should not poll for changes."""
+        return False
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        if self._attribute.update_type == UpdateType.POLL:
+            return self._data_manager.poll_data_update_coordinator.last_update_success
+
+        return True
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique, Home Assistant friendly identifier for this entity."""
+        return self._unique_id
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity, if any."""
+        return self._attribute.unit_of_measurement
+
+    @property
+    def icon(self) -> str:
+        """Icon to use in the frontend, if any."""
+        return self._attribute.icon
+
+    @property
+    def state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes.
+
+        Implemented by component base class. Convention for attribute names
+        is lowercase snake_case.
+        """
+        return {"friendly_name": self._attribute.friendly_name}
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return self._attribute.enabled_by_default
+
+    @callback
+    def _on_poll_data_updated(self) -> None:
+        self._update_state_data(
+            self._data_manager.poll_data_update_coordinator.data or {}
+        )
+
+    @callback
+    def _on_webhook_data_updated(self) -> None:
+        self._update_state_data(
+            self._data_manager.webhook_update_coordinator.data or {}
         )
 
-        return self._sleep_summary
+    def _update_state_data(self, data: MeasurementData) -> None:
+        """Update the state data."""
+        self._state_data = data.get(self._attribute.measurement)
+        self.schedule_update_ha_state()
```

Comment:
We have to use the async api within async context.
```suggestion
        self.async_write_ha_state()
```